### PR TITLE
Fix ProductNames wildcard processing 

### LINKED ISSUE #843: ScubaConfig.ClearConfiguration doesn't clear configuration
# 🐛 Summary #

After `[scubaconfig]::ResetInstance()` the old configuration is still present. The steps in .ClearConfiguration to loop through the keys and delete all keys in the hashtable seems to do something other than intended. It may be actually deleting object properties that aren't set on the base object. Either way, the keys still remain after .ClearConfiguration has been called. 

## To reproduce ##

```
using module ./PowerShell/ScubaGear/Modules/ScubaConfig/ScubaConfig.psm1

[ScubaConfig]::ResetInstance()
$cfg = [ScubaConfig]::GetInstance()
$cfg.Configuration = @{}
$cfg.Configuration.ProductNames = @("teams")
write-host "ProductNames: " $cfg.Configuration.ProductNames
write-host "Resetting instance"
[ScubaConfig]::ResetInstance()
write-host "ProductNames is still: " $cfg.Configuration.ProductNames
```
The output of this script is:
```
> ~/pstest/scubaconfigtest.ps1
ProductNames:  teams
Resetting instance
ProductNames is still:  teams
```
## Expected behavior ##

After calling `[scubaconfig]::ResetInstance()` the `.Configuration` hashtable should have no keys in it. The loop below doesn't actually delete the keys:
```
    hidden [void]ClearConfiguration(){
        Get-Member -InputObject ($this.Configuration) -Type properties |
          ForEach-Object { $this.Configuration.PSObject.Properties.Remove($_.name)}
    }
```
It should be easy to fix by deleting the hashtable keys similar to below:
```
    hidden [void]ClearConfiguration(){
        Get-Member -InputObject ($this.Configuration) -Type properties |
          ForEach-Object { $this.Configuration.Remove($_.name)}
    }
```
or by just completely overwriting .Configuration with an empty hashtable:
```
   $this.Configuration = @{}
```
Happy to take this one if it'd be helpful.

### LINKED ISSUE #884: ProductNames does not accept wildcard to override config file
# 🐛 Summary #

The `Invoke-Scuba` cmdlet takes the `-ProductNames` parameter as a list of baselines to be assessed.  It can take a list of products or a wildcard represented by `*` to mean all supported products.  However, if used with the `-ConfigFilePath` option, `-ProductNames` does not support the wildcard and generates an error.

## To reproduce ##

1. Create a valid configuration file for ScubaGear (use sample configs for ease of use)
2. Then run `Invoke-Scuba -ConfigFilePath config.yaml -ProductNames *`

## Expected behavior ##

Expected the wildcard to be accepted and converted into the full list of supported products.

## Any helpful log output or screenshots ##

Example invocation using a config file and product names wildcard
Screenshot 2024-02-01 at 8 45 13 AM

### PR DESCRIPTION
**Pull request by crutchfield**
#  #
  

## 🗣 Description ##

ProductNames does not accept wildcard to override config file.  This fixes the handling of of ProductNames in both command line and config file. 
Closes #884 
Closes #843 

## 💭 Motivation and context ##

This will allow users to get expected results when overriding the ProductsNames via command line.  

## 🧪 Testing ##

- Added unit test to check for ProductNames wildcard handling, all pass
- Fix unit test for ScubaConfig reset
- Ran InvokeSCuBA with configuration file and wildcard ProductNames override

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] Changes are sized such that they do not touch excessive number of files.
- [X] *All* future TODOs are captured in issues, which are referenced in code comments.
- [X] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [X] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [X] All relevant type-of-change labels added.
- [X] All relevant project fields are set.
- [X] All relevant repo and/or project documentation updated to reflect these changes.
- [X] Unit tests added/updated to cover PowerShell and Rego changes.
- [X] Functional tests added/updated to cover PowerShell and Rego changes.
- [X] All relevant functional tests passed.
- [X] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Review by Dylan-MITRE — COMMENTED** [review]
For issue #884 Invoke-Scuba -ConfigFilePath config.yaml -ProductNames * works now, the OPA issue still exists but it is being working on in another issue (#894).

For Issue #843 I ran the script that Thom wrote and it still doesnt seem like the [ScubaConfig]::ResetInstance() is fixed

reset

**Comment by crutchfield** [comment]
> For issue #884 Invoke-Scuba -ConfigFilePath config.yaml -ProductNames * works now, the OPA issue still exists but it is being working on in another issue (#894).
> 
> For Issue #843 I ran the script that Thom wrote and it still doesnt seem like the [ScubaConfig]::ResetInstance() is fixed
> 
> reset

Please include script referenced so I can evaluate.

**Comment by Dylan-MITRE** [comment]
> > For issue #884 Invoke-Scuba -ConfigFilePath config.yaml -ProductNames * works now, the OPA issue still exists but it is being working on in another issue (#894).
> > For Issue #843 I ran the script that Thom wrote and it still doesnt seem like the [ScubaConfig]::ResetInstance() is fixed
> > reset
> 
> Please include script referenced so I can evaluate.

https://github.com/cisagov/ScubaGear/issues/843#issue-2098658096

this one

**Comment by crutchfield** [comment]
> > > For issue #884 Invoke-Scuba -ConfigFilePath config.yaml -ProductNames * works now, the OPA issue still exists but it is being working on in another issue (#894).
> > > For Issue #843 I ran the script that Thom wrote and it still doesnt seem like the [ScubaConfig]::ResetInstance() is fixed
> > > reset
> > 
> > 
> > Please include script referenced so I can evaluate.
> 
> [#843 (comment)](https://github.com/cisagov/ScubaGear/issues/843#issue-2098658096)
> 
> this one



> > > For issue #884 Invoke-Scuba -ConfigFilePath config.yaml -ProductNames * works now, the OPA issue still exists but it is being working on in another issue (#894).
> > > For Issue #843 I ran the script that Thom wrote and it still doesnt seem like the [ScubaConfig]::ResetInstance() is fixed
> > > reset
> > 
> > 
> > Please include script referenced so I can evaluate.
> 
> [#843 (comment)](https://github.com/cisagov/ScubaGear/issues/843#issue-2098658096)
> 
> this one

The script is not used how the class was intended to be used.  But I made a change to accommodate if it is misused (setting config directly instead of 'loading' from a config file).  Good catch.

**Review by Dylan-MITRE — APPROVED** [review]

File: PowerShell/ScubaGear/Testing/Unit/PowerShell/ScubaConfig/ScubaConfigDelete.Tests.ps1 [hunk 1]
```diff
@@ -12,12 +12,21 @@ InModuleScope ScubaConfig {
                 $Result | Should -Be $true
             }
             It '6 Product names'{
+                $ScubaConfigTestFile = Join-Path -Path $PSScriptRoot -ChildPath config_test.yaml
+                $Result = [ScubaConfig]::GetInstance().LoadConfig($ScubaConfigTestFile)
+                $Result | Should -Be $true
                 [ScubaConfig]::GetInstance().Configuration.ProductNames | Should -HaveCount 6 -Because "$([ScubaConfig]::GetInstance().Configuration.ParameterNames)"
```
[the remark this answers is not in the document]
Comment (crutchfield) [reply PowerShell/ScubaGear/Testing/Unit/PowerShell/ScubaConfig/ScubaConfigDelete.Tests.ps1:18]: @schrolla  Updated config_test.yaml and config_test.json

**Comment by crutchfield** [comment]
@nanda-katikaneni PR is ready for merge to main

**Pull request merged**